bee-protocol/src/workers/peer/manager_res.rs

```diff
@@ -48,64 +55,95 @@ type PeerTuple = (Arc<Peer>, Option<(GossipSender, oneshot::Sender<()>)>);
 pub struct PeerManagerInner {
     peers: HashMap<PeerId, PeerTuple>,
     // This is needed to ensure message distribution fairness as iterating over a HashMap is random.
-    // TODO private
-    pub(crate) keys: Vec<PeerId>,
+    keys: Vec<PeerId>,
 }
 
 #[derive(Default)]
-// TODO private
-pub struct PeerManager(pub(crate) RwLock<PeerManagerInner>);
+pub struct PeerManager {
+    inner: RwLock<PeerManagerInner>,
+    counter: AtomicUsize,
+}
 
 impl PeerManager {
     pub(crate) fn new() -> Self {
         Self::default()
     }
 
     pub fn is_empty(&self) -> bool {
-        self.0.read().peers.is_empty()
+        self.inner.read().peers.is_empty()
     }
 
     // TODO find a way to only return a ref to the peer.
     pub fn get(&self, id: &PeerId) -> Option<impl std::ops::Deref<Target = PeerTuple> + '_> {
-        RwLockReadGuard::try_map(self.0.read(), |map| map.peers.get(id)).ok()
+        RwLockReadGuard::try_map(self.inner.read(), |map| map.peers.get(id)).ok()
     }
 
     pub fn get_mut(&self, id: &PeerId) -> Option<impl std::ops::DerefMut<Target = PeerTuple> + '_> {
-        RwLockWriteGuard::try_map(self.0.write(), |map| map.peers.get_mut(id)).ok()
+        RwLockWriteGuard::try_map(self.inner.write(), |map| map.peers.get_mut(id)).ok()
     }
 
     pub fn get_all(&self) -> Vec<Arc<Peer>> {
-        self.0.read().peers.iter().map(|(_, (peer, _))| peer).cloned().collect()
+        self.inner
+            .read()
+            .peers
+            .iter()
+            .map(|(_, (peer, _))| peer)
+            .cloned()
+            .collect()
     }
 
     pub(crate) fn add(&self, peer: Arc<Peer>) {
         debug!("Added peer {}.", peer.id());
-        let mut lock = self.0.write();
+        let mut lock = self.inner.write();
         lock.keys.push(*peer.id());
         lock.peers.insert(*peer.id(), (peer, None));
     }
 
     pub(crate) fn remove(&self, id: &PeerId) -> Option<PeerTuple> {
         debug!("Removed peer {}.", id);
-        let mut lock = self.0.write();
+        let mut lock = self.inner.write();
         lock.keys.retain(|peer_id| peer_id != id);
         lock.peers.remove(id)
     }
 
     pub(crate) fn for_each<F: Fn(&PeerId, &Peer)>(&self, f: F) {
-        self.0.read().peers.iter().for_each(|(id, (peer, _))| f(id, peer));
+        self.inner.read().peers.iter().for_each(|(id, (peer, _))| f(id, peer));
+    }
+
+    pub(crate) fn fair_find(&self, f: impl Fn(&Peer) -> bool) -> Option<PeerId> {
```

Comment:
Not necessarily related to that PR, but I think a little prose to explain why this `fair_find` method is needed could be helpful. Could we use an `IndexMap` here?

Isn't a better method name: `select_peer_randomly` or something?

Reply:
I think I'm not the right person to answer this as I didn't write the code that inspired this method. 

I also have a couple questions about the `Vec<PeerId>`. The iteration order in `HashMap::keys` is not deterministic between runs but it is consistent if the hashmap has not changed, why is this not enough?

Regarding the name. I favor `find` against `select` because of [`Iterator::find`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find). About the `randomly` instead of `fairly` I'd say this is not a random traverse, it is consistent if the hashmap does not change.

Reply:
> but it is consistent if the hashmap has not changed, why is this not enough?

Precisely because the maps does change a lot.

Reply:
@thibault-martinez but if the map changes a lot, the vec changes a lot so what difference does it make?

Reply:
Sure but the order is not completely shuffled, it keeps being fair for remaining peers

Reply:
but removing a key from a hashmap doesn't change the iteration order for the other keys.

Reply:
Right, I now do better understand why the method was written the way it is, but - coming back to my original point - in order to reduce confusion I think the code is "odd" (with the atomic) enough to add a little bit of explanation. Instead of that rotating "pointer" (modulus counter), one could also go with a ring data structure here, which feels like it would better encode the situation we have here.

Reply:
The point of the counter is to keep the state of where you stopped cycling for a future iteration. See Discord protocol, Christian will try something.